File: sdks/python/opencomputer/pty.py
"""PTY terminal sessions inside a sandbox."""

from __future__ import annotations

import asyncio
from dataclasses import dataclass
from typing import Callable

import httpx
import websockets


@dataclass
class PtySession:
    """An active PTY terminal session."""

    session_id: str
    sandbox_id: str
    _ws: websockets.WebSocketClientProtocol | None = None
    _read_task: asyncio.Task | None = None

    async def send(self, data: str | bytes) -> None:
        """Send input to the terminal."""
        if self._ws is None:
            raise RuntimeError("PTY session not connected")
        payload = data if isinstance(data, bytes) else data.encode()
        await self._ws.send(payload)

    async def recv(self) -> bytes:
        """Receive output from the terminal."""
        if self._ws is None:
            raise RuntimeError("PTY session not connected")
        data = await self._ws.recv()
        return data if isinstance(data, bytes) else data.encode()

    async def close(self) -> None:
        """Close the PTY session."""
        if self._read_task and not self._read_task.done():
            self._read_task.cancel()
        if self._ws:
            await self._ws.close()


@dataclass
class Pty:
    """PTY terminal session manager for a sandbox."""

    _client: httpx.AsyncClient
    _sandbox_id: str
    _api_url: str
    _api_key: str

    async def create(
        self,
        cols: int = 80,
        rows: int = 24,
        on_output: Callable[[bytes], None] | None = None,
    ) -> PtySession:
        """Create a new PTY session and connect via WebSocket."""
        # Create session via REST
        resp = await self._client.post(
            f"/sandboxes/{self._sandbox_id}/pty",
            json={"cols": cols, "rows": rows},
        )
        resp.raise_for_status()
        data = resp.json()
        session_id = data["sessionID"]

        # Connect via WebSocket
        ws_url = self._api_url.replace("http://", "ws://").replace("https://", "wss://")
        ws_url = f"{ws_url}/sandboxes/{self._sandbox_id}/pty/{session_id}"

        headers = {}
        if self._api_key:
            headers["X-API-Key"] = self._api_key

        ws = await websockets.connect(ws_url, additional_headers=headers)

        session = PtySession(
            session_id=session_id,
            sandbox_id=self._sandbox_id,
            _ws=ws,
        )

        if on_output:
            async def _reader() -> None:
                try:
                    async for msg in ws:
                        output = msg if isinstance(msg, bytes) else msg.encode()
                        on_output(output)
                except websockets.ConnectionClosed:
                    pass

            session._read_task = asyncio.create_task(_reader())

        return session

File: sdks/typescript/src/pty.ts
export interface PtyOpts {
  cols?: number;
  rows?: number;
  onOutput?: (data: Uint8Array) => void;
}

export interface PtySession {
  sessionId: string;
  send(data: string | Uint8Array): void;
  close(): void;
}

export class Pty {
  constructor(
    private apiUrl: string,
    private apiKey: string,
    private sandboxId: string,
    private token: string = "",
  ) {}

  private get headers(): Record<string, string> {
    const h: Record<string, string> = { "Content-Type": "application/json" };
    if (this.token) {
      h["Authorization"] = `Bearer ${this.token}`;
    } else if (this.apiKey) {
      h["X-API-Key"] = this.apiKey;
    }
    return h;
  }

  async create(opts: PtyOpts = {}): Promise<PtySession> {
    // Create session via REST
    const resp = await fetch(`${this.apiUrl}/sandboxes/${this.sandboxId}/pty`, {
      method: "POST",
      headers: this.headers,
      body: JSON.stringify({
        cols: opts.cols ?? 80,
        rows: opts.rows ?? 24,
      }),
    });

    if (!resp.ok) {
      throw new Error(`Failed to create PTY: ${resp.status}`);
    }

    const data = await resp.json();
    const sessionId: string = data.sessionID;

    // Connect via WebSocket
    const wsUrl = this.apiUrl
      .replace("http://", "ws://")
      .replace("https://", "wss://");
    const tokenParam = this.token ? `?token=${encodeURIComponent(this.token)}` : "";
    const wsEndpoint = `${wsUrl}/sandboxes/${this.sandboxId}/pty/${sessionId}${tokenParam}`;

    const ws = new WebSocket(wsEndpoint);
    ws.binaryType = "arraybuffer";

    if (opts.onOutput) {
      const onOutput = opts.onOutput;
      ws.onmessage = (event) => {
        const data = event.data instanceof ArrayBuffer
          ? new Uint8Array(event.data)
          : new TextEncoder().encode(event.data as string);
        onOutput(data);
      };
    }

    return {
      sessionId,
      send(data: string | Uint8Array): void {
        if (typeof data === "string") {
          ws.send(data);
        } else {
          ws.send(data);
        }
      },
      close(): void {
        ws.close();
      },
    };
  }
}
